src/main/scala/cromwell/engine/backend/io/package.scala

```diff
@@ -0,0 +1,58 @@
+package cromwell.engine.backend
+
+import java.nio.file._
+
+import cromwell.engine.backend.io.filesystem.gcs.{ContentTypeOption, GcsFileSystem, GcsFileSystemProvider, NioGcsPath}
+
+import scala.util.{Success, Try}
+
+package object io {
+  object PathString {
+    val defaultGCSFileSystem = GcsFileSystem.defaultGcsFileSystem
+    val defaultFileSystem = FileSystems.getDefault
+    private val defaultFileSystemsList = List(defaultGCSFileSystem, defaultFileSystem)
+
+    implicit class UriString(val str: String) extends AnyVal {
+
+      def isGcsUrl: Boolean = str.startsWith("gs://")
+      def isUriWithProtocol: Boolean = "^[a-z]+://".r.findFirstIn(str).nonEmpty
+
+      def toPath(implicit fss: List[FileSystem] = defaultFileSystemsList): Path = {
+        fss.toStream map { fs => Try(fs.getPath(str)) } collectFirst { case Success(p) => p } getOrElse {
+          throw new IllegalArgumentException(s"Could not find suitable filesystem to parse $str")
+        }
+      }
+
+      def toDirectory(implicit fss: List[FileSystem] = defaultFileSystemsList): Path = {
+        fss.toStream collect {
+          case fs: GcsFileSystem => Try(fs.getPathAsDirectory(str))
+          case fs => Try(fs.getPath(str))
+        } collectFirst { case Success(p) => p } getOrElse {
+          throw new IllegalArgumentException(s"Could not find suitable filesystem to parse $str")
+        }
+      }
+
+      def toAbsolutePath(implicit fss: List[FileSystem] = defaultFileSystemsList): Path = {
+        fss.toStream map { fs => Try(fs.getPath(str).toAbsolutePath) } collectFirst { case Success(p) => p } getOrElse {
+          throw new IllegalArgumentException(s"Could not find suitable filesystem to parse $str")
+        }
+      }
+    }
+
+    implicit class PathEnhanced(val path: Path) extends AnyVal {
+      import better.files._
+
+      def hash = path match {
+        case gcs: NioGcsPath => gcs.getFileSystem.provider().asInstanceOf[GcsFileSystemProvider].crc32cHash(gcs)
+        case _ => path.md5
```

Comment:
Argh this is not going to help my case for implicits :s
This is actually an implicit method in better.files._ that reads the file and hash the content (which by the way will need to be change anyway since it loads everything in memory and blows up for big files...)